src/cata_tiles.h

```diff
@@ -25,9 +25,37 @@ class JsonObject;
 struct visibility_variables;
 
 /** Structures */
+struct sprite_variation {
+    int weight;
+    std::vector<int> sprite_ids;
+
+    sprite_variation()
+    {
+        weight = 1;
+        sprite_ids.clear();
+    }
+
+    sprite_variation(int id)
+    {
+        weight = 1;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+
+    sprite_variation(int w, int id)
+    {
+        weight = w;
+        sprite_ids.clear();
+        sprite_ids.push_back(id);
+    }
+};
+
 struct tile_type
 {
-    std::vector<int> fg, bg;
+    // first entry of each of these vectors is not a real sprite_variation
+    // but a bookkeeping list instead, used for reducing the task of selecting
+    // a random variation from O(N) to O(1)
```

Comment:
the special part only appears in one such struct, out of a list of structs. if I moved it out, it would have to go adjacent to the list of structs, adding something like `vector<int> fg_pick, bg_pick` to tile_type. I considered that, but didn't like that concept.

using a class like weighted_list would resolve this problem, but as I mentioned elsewhere I'd have to improve that class first, and I didn't want to build such a class entirely from scratch here.